microceph/api/remote.go

```diff
@@ -0,0 +1,160 @@
+package api
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/canonical/lxd/lxd/response"
+	"github.com/canonical/lxd/shared/logger"
+	"github.com/canonical/microceph/microceph/api/types"
+	"github.com/canonical/microceph/microceph/ceph"
+	"github.com/canonical/microceph/microceph/client"
+	"github.com/canonical/microceph/microceph/constants"
+	"github.com/canonical/microceph/microceph/database"
+	"github.com/canonical/microceph/microceph/interfaces"
+	"github.com/canonical/microcluster/v2/rest"
+	"github.com/canonical/microcluster/v2/state"
+	"github.com/gorilla/mux"
+)
+
+// remoteCmd is the top level remote endpoint.
+var remoteCmd = rest.Endpoint{
+	Path: "client/remotes",
+	Get:  rest.EndpointAction{Handler: cmdRemoteGet, ProxyTarget: false},
+}
+
+// remoteNameCmd endpoint is for operations on specific remotes.
+var remoteNameCmd = rest.Endpoint{
+	Path:   "client/remotes/{name}",
+	Put:    rest.EndpointAction{Handler: cmdRemotePut, ProxyTarget: false},
+	Get:    rest.EndpointAction{Handler: cmdRemoteGet, ProxyTarget: false},
+	Delete: rest.EndpointAction{Handler: cmdRemoteDelete, ProxyTarget: false},
+}
+
+// cmdRemotePut is handler for adding remote records to MicroCeph.
+// This also triggers the $cluster file generation for all MicroCeph hosts.
+func cmdRemotePut(state state.State, r *http.Request) response.Response {
+	var req types.RemoteImportRequest
+
+	err := json.NewDecoder(r.Body).Decode(&req)
+	if err != nil {
+		return response.InternalError(err)
+	}
+
+	err = renderConfAndKeyringFiles(req.Name, req.LocalName, req.Config)
+	if err != nil {
+		return response.InternalError(fmt.Errorf("couldn't render files: %w", err))
+	}
+
+	if !req.RenderOnly {
+		logger.Infof("REM: Sending remote(%s) info to cluster members.", req.Name)
+
+		// Asynchronously persist this on db and send request to other cluster members.
+		go func() {
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second*120)
+			defer cancel()
+
+			err := database.PersistRemoteDb(ctx, interfaces.CephState{State: state}, req)
+			if err != nil {
+				logger.Errorf("REM: failed to persiste remote: %s", err.Error())
+			}
+
+			// Send render only request to remaining cluster members.
+			req.RenderOnly = true
+			err = client.SendRemoteImportToClusterMembers(ctx, state, req)
```

Comment:
How would the remote part know that an error occurred? (i.e: If the database couldn't persist the information). Shouldn't an error be sent in that case?

Reply:
Since this is the first implementation, we are sticking to the dual import mechanism. I would want to have a to-fro mechanism between the clusters but at this point I feel we may be implementing something that will be replaced by the microcluster implementation.